#!/bin/bash
#$ -cwd
#$ -V
#$ -N ldpred_gibbs
#$ -l h_data=16G,h_rt=12:00:00
#$ -t 1-18:1

# note:
# -o: what directory to write any printed output or error messages to
# -t: makes it a job array. change 1:14:1 to however many traits you are running

source /etc/profile     # so module command is recognized
module load anaconda
source activate ldpred

# The above is an example for running this code in UCLA Hoffman2 Job submission system, the user may change the code for different job submission systems. 

# PHENO: get the phenotype names from a text file with one name per line
# the phenotype names in the lines for MT estimations are the trait names followed by the number of phenotypes in the set, and the first letter of all other phenotypes in the set. For example, if the three phenotypes in a set of analysis is armfat_percent, height and pulse, the final result produced in the analysis for armfat_percent is armfat_percent_3hp. Example see MT_list_example.txt.
# the phenotype names in the lines for GWAS summary statistics are the trait names listed in pheno_name_list that correspond to GWAS summary statistics names. Example see Pheno_list_example.txt.

# N: The number of individuals (N) participated in the study for each of the snp ### TODO: ASK Camille for n.txt.

# GFILE: One genotype file with at least 1000 unrelated individuals with the same ancestry make-up as the individuals for which summary statistics datasets are obtained from. 

# COORDFILE: The HDF5 file generated by the previous step coord.
# LDF_FILE:  ???
# OUTFILE: This step generates two files, a LD file with LD information for the given LD radius, and the re-weighted effect estimates. 

# get the phenotype name from a text file with one name per line
PHENO="`sed -n ${SGE_TASK_ID}p PATH_TO_GWAS_list/pheno_name_list.txt`"

N="`sed -n ${SGE_TASK_ID}p n.txt`"

GFILE="Genotype_file (directory and file name without suffix .bed/.bim/.fam)"

COORDFILE="PATH_TO_COORD_FILE/ldpred_coord_${PHENO}"

LDF_FILE="ldf_${PHENO}"

OUTFILE="PATH_TO_OUT_FILE/ldpred_gibbs_${PHENO}"


# --ldr: LD radius. We used 150 in our settings. The user could change for personal need.
# --f: Fraction of causal variants used in the Gibbs sampler. Here, as we assume an infinitesimal model where each variant has an effect on all of the traits, --f is 1.



echo $PHENO
date


ldpred gibbs --cf ${COORDFILE} --ldr 70 --ldf ${LDF_FILE} --out ${OUTFILE} --N ${N} --f 1


echo $PHENO # makes it easy to check which trait unfinished
date





