examples in format 
========
before
========
after
========

MAF needed for MTAG
	generate MAF from bfile using plink with:
		generateMAF.sh

		.frq file:
		 CHR           SNP   A1   A2          MAF  NCHROBS
		1      1:801536    G    T      0.01481   300000
   		1      1:835499    G    A        0.241   300000
   		1      1:838555    A    C       0.2462   300000
   		1      1:840753    C    T       0.4003   300000
   		1      1:846864    C    G       0.1986   300000
		

	add MAF to plink gwas with:
		addMAF_to_plinkout.py
		==================================================================================================================
		chr     bpos    snpid   a2      a1      n       beta    se      pval    z
		1       801536  1:801536        T       G       147448  -0.00202558     0.0112628       0.857273        -0.179846929716
		1       835499  1:835499        A       G       147448  -0.00182929     0.00317687      0.564742        -0.575815189164
		1       838555  1:838555        C       A       147448  -0.00309543     0.00315503      0.32654 -0.981109529862
		=======================================================================================
		chr     bpos    snpid   a2      a1      n       beta    se      pval    z       MAF
		1       801536  1:801536        T       G       147448  -0.00202558     0.0112628       0.857273        -0.179846929716 0.01481
		1       835499  1:835499        A       G       147448  -0.00182929     0.00317687      0.564742        -0.575815189164 0.241
		1       838555  1:838555        C       A       147448  -0.00309543     0.00315503      0.32654 -0.981109529862 0.2462
		==================================================================================================
A2 allele needed by LDpred, but not in plink sumstats:

	addA2allele_to_new_plinkout.py -- adds A2 allele to gwas summary statistics generated from plink
	=================================================================================================================================
	CHR           SNP         BP   A1       TEST    NMISS       BETA       SE      L95      U95         STAT            P
   	1      1:801536     801536    G        ADD   196594  0.0003117 0.001669 -0.00296 0.003583       0.1868       0.8518
   	1      1:835499     835499    G        ADD   196594  0.0006431  0.00167 -0.002629 0.003916       0.3852       0.7001
   	1      1:838555     838555    A        ADD   196594 -0.0007357 0.001669 -0.004007 0.002536      -0.4407       0.6594
	=============================================================================================================================
	CHR           SNP         BP   A1       TEST    NMISS       BETA       SE      L95      U95         STAT            P  A2
   	1      1:801536     801536    G        ADD   196594  0.0003117 0.001669 -0.00296 0.003583       0.1868       0.8518  T
   	1      1:835499     835499    G        ADD   196594  0.0006431  0.00167 -0.002629 0.003916       0.3852       0.7001 A
  	1      1:838555     838555    A        ADD   196594 -0.0007357 0.001669 -0.004007 0.002536      -0.4407       0.6594 C
	=============================================================================================================================

for adding MAF, CHR, BP, A1, A2 to Neale group:
	reformat_neale.py
	==============================================================================================================================
	variant rsid    nCompleteSamples        AC      ytx     beta    se      tstat   pval
	5:43888254:C:T  rs13184706      317756  1.16108e+04     -3.90624e+01    4.62875e-03     9.41073e-03     4.91859e-01     6.22820e-01
	5:43888493:C:T  rs58824264      317756  2.23066e+03     -2.27979e+01    -5.40375e-03    2.12034e-02     -2.54853e-01    7.98837e-01
	5:43888556:T:C  rs72762387      317756  1.56871e+04     -7.96053e+01    5.77564e-03     8.11979e-03     7.11303e-01     4.76897e-01
	===========================================================================================================================
	variant rsid    nCompleteSamples        AC      ytx     beta    se      tstat   pval    chr     bpos    a1      a2      MAF
	5:43888254:C:T  rs13184706      317756  1.16108e+04     -3.90624e+01    4.62875e-03     9.41073e-03     4.91859e-01     6.22820e-01     5       43888254        C       T       -6.14660305392e-05
	5:43888493:C:T  rs58824264      317756  2.23066e+03     -2.27979e+01    -5.40375e-03    2.12034e-02     -2.54853e-01    7.98837e-01     5       43888493        C       T       -3.58732801269e-05
	5:43888556:T:C  rs72762387      317756  1.56871e+04     -7.96053e+01    5.77564e-03     8.11979e-03     7.11303e-01     4.76897e-01     5       43888556        T       C       -0.000125261678772
	=====================================================================================================================================

SNP identification method for reference panel must match SNP identification for sumstats --i.e. if one has rsID and the other is chr:bp, must convert one to the other:
Additionally, mtag snp filter is in rsID -- to intersect with filter4 snps, must have rsID in filter 4
	for converting chr:pos to rsID: 
		for plink sumstats (after adding MAF):
			chrpos_to_rsID.py
			===============================================================================================================================
			chr     bpos    snpid   a2      a1      n       beta    se      pval    z       MAF
			1       801536  1:801536        T       G       147448  -0.00202558     0.0112628       0.857273        -0.179846929716 0.01481
			1       835499  1:835499        A       G       147448  -0.00182929     0.00317687      0.564742        -0.575815189164 0.241
			1       838555  1:838555        C       A       147448  -0.00309543     0.00315503      0.32654 -0.981109529862 0.2462
			=================================================================================================================================
			chr     bpos    snpid   a2      a1      n       beta    se      pval    z       MAF     rsID
			1       849998  1:849998        G       A       147448  0.0100705       0.00354413      0.00449119      2.84145897583   0.179   rs13303222
			1       850780  1:850780        T       C       147448  0.00647477      0.00278447      0.0200563       2.32531505098   0.3961  rs6657440
			1       880238  1:880238        G       A       147448  -0.00228944     0.00603599      0.704467        -0.379298176438 0.05346 rs3748592	
			===================================================================================================================================

		for plink bim files:
			chrpos_to_rsID_filter.py
			==================================================
			1       1:801536        0       801536  G       T
			1       1:835499        0       835499  G       A
			1       1:838555        0       838555  A       C
			==================================================
			1       1:801536        0       801536  G       T       rs79373928
			1       1:835499        0       835499  G       A       rs4422948
			1       1:838555        0       838555  A       C       rs4970383

			
	for adding adding snpid from chr:pos:
		for adding snpid to reformatted neale:
			addsnpid_neale.py
		================================================================
		variant rsid    nCompleteSamples        AC      ytx     beta    se      tstat   pval    chr     bpos    a1      a2      MAF
		5:43888254:C:T  rs13184706      317756  1.16108e+04     -3.90624e+01    4.62875e-03     9.41073e-03     4.91859e-01     6.22820e-01     5       43888254        C       T       -6.14660305392e-05
		5:43888493:C:T  rs58824264      317756  2.23066e+03     -2.27979e+01    -5.40375e-03    2.12034e-02     -2.54853e-01    7.98837e-01     5       43888493        C       T       -3.58732801269e-05
		5:43888556:T:C  rs72762387      317756  1.56871e+04     -7.96053e+01    5.77564e-03     8.11979e-03     7.11303e-01     4.76897e-01     5       43888556        T       C       -0.000125261678772
		======================================================================
		variant rsid    nCompleteSamples        AC      ytx     beta    se      tstat   pval    chr     bpos    a2      a1      MAF     snpid
		5:43888254:C:T  rs13184706      317756  1.16108e+04     -3.90624e+01    4.62875e-03     9.41073e-03     4.91859e-01     6.22820e-01     5       43888254        C       T       0.0182699933282 5:43888254
		5:43888493:C:T  rs58824264      317756  2.23066e+03     -2.27979e+01    -5.40375e-03    2.12034e-02     -2.54853e-01    7.98837e-01     5       43888493        C       T       0.00351002026712        5:43888493
		5:43888556:T:C  rs72762387      317756  1.56871e+04     -7.96053e+01    5.77564e-03     8.11979e-03     7.11303e-01     4.76897e-01     5       43888556        T       C       0.0246841916439 5:43888556
		
		for adding snpid to reformatted neale, filtered (filter also removes AC, ytx, and tstat columns)
			addsnpid_to_filtered_neale.py
		====================================================================
		variant SNP     NMISS   BETA    SE      STAT    P       CHR     BP      A1      A2      MAF
		6:2097686:C:T   rs4959631       331249  3.77597e-04     1.85662e-03     2.03379e-01     8.38839e-01     6       2097686 C       T       -0.00980926734873
		6:2198327:G:A   rs234944        331249  5.83682e-04     1.90613e-03     3.06213e-01     7.59443e-01     6       2198327 G       A       -0.0106534812181
		6:2204274:A:G   rs234954        331249  -1.88740e-04    1.89015e-03     -9.98546e-02    9.20460e-01     6       2204274 A       G       -0.0106378283406
		===========================================================================
		variant SNP     NMISS   BETA    SE      STAT    P       CHR     BP      A1      A2      MAF     snpid
		6:2097686:C:T   rs4959631       317756  -1.49319e-03    2.51567e-03     -5.93555e-01    5.52810e-01     6       2097686 C       T       -0.00486230000378       6:2097686
		6:2198327:G:A   rs234944        317756  -1.23543e-03    2.58279e-03     -4.78329e-01    6.32416e-01     6       2198327 G       A       -0.00525938141215       6:2198327
		6:2204274:A:G   rs234954        317756  -1.94899e-04    2.56148e-03     -7.60883e-02    9.39349e-01     6       2204274 A       G       -0.00484440891753       6:2204274		

		for adding snpid to MT results on neale:
			addsnpid_to_neale_MT.py
		==========================================================================================================
		SNP     variant CHR     BP      A1      NMISS   BETA    SE      STAT    P       A2      MAF     BETA_MTAG       BETA_GMMa       BETA_GMMb       BETA_GMMc       BETA_GMMd       PVAL_MTAG       PVAL_GMMa       PVAL_GMMb       PVAL_GMMc       PVAL_GMMd
		rs4959631       6:2097686:C:T   6       2097686 C       331249  3.77597e-04     1.85662e-03     2.03379e-01     8.38839e-01     T       -0.00980926734873       -0.00034825434099354925 0.0021570679346580805   6.06520628853741e-06    0.0003937874430744535   0.00045040279596526796  0.8512103133876512      0.24530646867325934     0.997393475321728       0.8320297167717253      0.8083207961240874
		rs234944        6:2198327:G:A   6       2198327 G       331249  5.83682e-04     1.90613e-03     3.06213e-01     7.59443e-01     A       -0.0106534812181        -3.629293176122617e-05  0.0021518890431864893   1.1290914229823107e-05  0.0005720017986393095   0.0006154373840799336   0.9848091051666376      0.2589269873921144      0.9952737780431927      0.7641120093604042      0.7467916345342904
		rs234954        6:2204274:A:G   6       2204274 A       331249  -1.88740e-04    1.89015e-03     -9.98546e-02    9.20460e-01     G       -0.0106378283406        -0.0010067584475043892  0.0017201298193943422   -5.933500421365963e-06  -0.00011346934952426728 -3.147722840371572e-05  0.5942868543796396      0.3627964730609986      0.9974953095774298      0.9521302053279476      0.9867132059716075
		======================================================================================================
		SNP     variant CHR     BP      A1      NMISS   BETA    SE      STAT    P       A2      MAF     BETA_MTAG       BETA_GMMa       BETA_GMMb       BETA_GMMc       BETA_GMMd       PVAL_MTAG       PVAL_GMMa       PVAL_GMMb       PVAL_GMMc       PVAL_GMMd       snpid
		rs4959631       6:2097686:C:T   6       2097686 C       331249  3.77597e-04     1.85662e-03     2.03379e-01     8.38839e-01     T       -0.00980926734873       -0.00034825434099354925 0.0021570679346580805   6.06520628853741e-06    0.0003937874430744535   0.00045040279596526796  0.8512103133876512      0.24530646867325934     0.997393475321728       0.8320297167717253      0.8083207961240874      6:2097686
		rs234944        6:2198327:G:A   6       2198327 G       331249  5.83682e-04     1.90613e-03     3.06213e-01     7.59443e-01     A       -0.0106534812181        -3.629293176122617e-05  0.0021518890431864893   1.1290914229823107e-05  0.0005720017986393095   0.0006154373840799336   0.9848091051666376      0.2589269873921144      0.9952737780431927      0.7641120093604042      0.7467916345342904      6:2198327
		rs234954        6:2204274:A:G   6       2204274 A       331249  -1.88740e-04    1.89015e-03     -9.98546e-02    9.20460e-01     G       -0.0106378283406        -0.0010067584475043892  0.0017201298193943422   -5.933500421365963e-06  -0.00011346934952426728 -3.147722840371572e-05  0.5942868543796396      0.3627964730609986      0.9974953095774298      0.9521302053279476      0.9867132059716075      6:2204274
		=========================================================================================================

LDpred cannot handle missing values ("NA") from phenotype files:
	remove individuals with "NA" from plink pheno files: 
		remove_NA.py


find intersection between two snp filters:
	snp_filter_intersection.py

find intersection and fraction of overlap between plink fam files:
	validation_intersection.py

filter snps from sumstats containing rsID using mtag filter:
	filter_snps_set.py

filter snps from plink sumstats containing only chr:bp using intersection between filter4, mtag snp list:
	filter_snps_set_plink.py
